foglamp/FogLAMP
# FOGL-2471 - Add certificate based authentication to FogLAMP API

Steps to test:
1) sudo make clean. make. scripts/foglamp reset.

2) Start foglamp. On Configuration, select "authentication - mandatory". Save. Stop server. Restart.

3) Create a local CA:

        cd data/etc/certs
        openssl genrsa -out ca.key 4096
        openssl req -new -x509 -days 365 -key ca.key -out ca.cert -subj '/C=US/CN=MY-CA'

   **Must:** In "CN" value, it must NOT be "foglamp" as it is already taken by foglamp.cert. Should we decide to use foglamp.cert as CA, it will create conflicts later.

4) Create a user certificate:
   
         openssl genrsa -out user.key 1024
         openssl req -new -key user.key -out user.csr -subj '/C=IN/CN=user'

   **Must:** In "CN" value, provide an existing username. For testing, it can be "user" which is an existing user.

5) Sign user certificate with CA certificate:

        openssl x509 -req -days 365 -in user.csr -CA ca.cert -CAkey ca.key -set_serial 01 -out user.cert

6) Now we need to login to generate a token for an user session based on user certificate. Start postman. In a tab, select "POST http://localhost:8081/foglamp/login". In Body, select option "raw". Choose "Text" from dropdown. Now copy paste contents of user.cert in Body. Press "Send". Kind of below response appears:

  ```
    {
       "admin": false,
       "token": "eyJ0eXAiOiJKV1QiLCJhbGciOiJIUzI1NiJ9.eyJleHAiOjE1NTE4MDI2NTksInVpZCI6Mn0.UcjDhR47Bj1MLNSrrW7tvdBqJUM2QNV-P8nY-ULinY8",
       "uid": 2,
       "message": "Logged in successfully"
    }
   ```

   With curl:

`curl -T data/etc/certs/user.cert -X POST http://localhost:8081/foglamp/login -k | jq
`

7) In another postman tab, select "GET http://localhost:8081/foglamp/schedule". In Headers, add "Authorization" with value of token above. Press "Send". A list of schedule appears.

   With curl:

`curl -H "Authorization:eyJ0eXAiOiJKV1QiLCJhbGciOiJIUzI1NiJ9.eyJ1aWQiOjIsImV4cCI6MTU1MTgwNjkzN30.LtMNZZrKMJNCz8cL3DB84xlAFstfIKvfubdok_XKyFI" -X GET http://localhost:8081/foglamp/schedule -k | jq
`

8) Try step 6) with incorrect client certificate, message appears like,

```
401: Authentication failed
```

TODO (separately): 

1) GUI page to generate client certificate.
2) GUI login page to add client certificate contents.
3) GUI page to revoke client certificate including expiry
4) Mechanism to revoke client certificate including expiry

[FOGL-2557](https://scaledb.atlassian.net/browse/FOGL-2557) has been created for 1, 2, 3 above.
